include/zephyr/bluetooth/classic/hid_device.h

```diff
@@ -0,0 +1,185 @@
+/** @file
+ *  @brief HID Device Protocol handling.
+ */
+
+/*
+ * Copyright 2024 Xiaomi Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_BLUETOOTH_HID_DEVICE_H_
+#define ZEPHYR_INCLUDE_BLUETOOTH_HID_DEVICE_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <zephyr/bluetooth/classic/sdp.h>
+#include <zephyr/bluetooth/hci.h>
+#include <zephyr/bluetooth/l2cap.h>
+
+#define BT_HID_MAX_MTU         64
+#define BT_HID_REPORT_DATA_LEN 64
+
+/** @brief HID report types in get, set, data */
+enum {
+	BT_HID_REPORT_TYPE_OTHER = 0,
+	BT_HID_REPORT_TYPE_INPUT,
+	BT_HID_REPORT_TYPE_OUTPUT,
+	BT_HID_REPORT_TYPE_FEATURE,
+};
+
+/** @brief HID handshake response code */
+enum {
+	BT_HID_HANDSHAKE_RSP_SUCCESS = 0,
+	BT_HID_HANDSHAKE_RSP_NOT_READY,
+	BT_HID_HANDSHAKE_RSP_ERR_INVALID_REP_ID,
+	BT_HID_HANDSHAKE_RSP_ERR_UNSUPPORTED_REQ,
+	BT_HID_HANDSHAKE_RSP_ERR_INVALID_PARAM,
+	BT_HID_HANDSHAKE_RSP_ERR_UNKNOWN = 14,
+	BT_HID_HANDSHAKE_RSP_ERR_FATAL,
+};
+
+/** @brief HID protocol parameters for set protocal */
+enum {
+	BT_HID_PROTOCOL_BOOT_MODE = 0,
+	BT_HID_PROTOCOL_REPORT_MODE,
+};
+
+/** @brief HID event type */
+typedef enum bt_hid_event {
+	BT_HID_EVENT_GET_REPORT = 0,
+	BT_HID_EVENT_SET_REPORT,
+	BT_HID_EVENT_GET_PROTOCOL,
+	BT_HID_EVENT_SET_PROTOCOL,
+	BT_HID_EVENT_INTR_DATA,
+	BT_HID_EVENT_UNPLUG,
+	BT_HID_EVENT_SUSPEND,
+	BT_HID_EVENT_EXIT_SUSPEND,
```

Comment:
Compared to custom events, I prefer to use spec defined types.
Custom types are similar to those defined by specs, with a greater number than those defined by specs. And it is not as clear as described in the spec. When calling back, they use the same callback function. Some custom events require `uint8_t * data` and `uint16_t len`, while others do not. This undoubtedly increases the learning cost and difficulty for users.

Is it possible to separate the callback function of the interrupt channel from that of the control channel? For control channels, one callback is used for one control message/event.

Reply:
> Compared to custom events, I prefer to use spec defined types. Custom types are similar to those defined by specs, with a greater number than those defined by specs. And it is not as clear as described in the spec. When calling back, they use the same callback function. Some custom events require `uint8_t * data` and `uint16_t len`, while others do not. This undoubtedly increases the learning cost and difficulty for users.

maybe it is a handsome  way to define a event avoid confusion with protocol stack

such get_report would decode wrap, and report vaule wihtout hid header.

typedef enum bt_hid_event {
	BT_HID_EVENT_GET_REPORT = 0,
	BT_HID_EVENT_SET_REPORT,
	BT_HID_EVENT_GET_PROTOCOL,
	BT_HID_EVENT_SET_PROTOCOL,
	BT_HID_EVENT_INTR_DATA,
	BT_HID_EVENT_UNPLUG,
	BT_HID_EVENT_SUSPEND,
	BT_HID_EVENT_EXIT_SUSPEND,
} bt_hid_event_t; 

/* Define the HID transaction types
 */
#define BT_HID_TYPE_HANDSHAKE    0x00
#define BT_HID_TYPE_CONTROL      0x01
#define BT_HID_TYPE_GET_REPORT   0x04
#define BT_HID_TYPE_SET_REPORT   0x05
#define BT_HID_TYPE_GET_PROTOCOL 0x06
#define BT_HID_TYPE_SET_PROTOCOL 0x07
#define BT_HID_TYPE_DATA         0x0a

/* Parameters for Control
 */
#define BT_HID_CONTROL_SUSPEND              0x03
#define BT_HID_CONTROL_EXIT_SUSPEND         0x04
#define BT_HID_CONTROL_VIRTUAL_CABLE_UNPLUG 0x05

Reply:
> Is it possible to separate the callback function of the interrupt channel from that of the control channel? For control channels, one callback is used for one control message/event.

Do you mean we specify the callback function as  below?

void (*set_report_req)(...)
void (*get_report_req)(...)
void (*set_protocol_req)(...)
void (*get_protocol_req)(...)
void (*intr_data_req)(...)
......

maye it's more comfortable